.github/workflows/test.yaml

```
@@ -76,9 +76,19 @@ jobs:
       - name: Lint OpenAPI specification file
         run: docs/openapi/openapi-lint
 
+  check-no-unwanted-strings:
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Check unwanted tags
+        run: |
+          grep -rEn --include='*.feature' '@(testing|debug)' . && { echo 'ERROR: Unwanted tags found.'; exit 1; } || echo 'No unwanted tags found in .feature files.'
```

`grep` returns exit code `1` (failure) if there is no match, and `-v` (which inverts `grep`’s match) wouldn’t give the desired result either, so I get why you were tempted to use `<cmd> && exit 1 || <success>`[^1].

However, since `grep` already prints the position of matches by default and GitHub says which job failed, we can simply negate `grep`’s status code and get rid of the `echo`s to get the desired result:

```suggestion
        run: ! grep -rEn --include='*.feature' '@(testing|debug)' .
```

[^1]: FYI, you could have also used `:` instead of `echo <success_message>`, which is a Bash builtin which returns `0` (see [Bourne Shell Builtins (Bash Reference Manual)](https://www.gnu.org/software/bash/manual/html_node/Bourne-Shell-Builtins.html#index-_003a)).

---

Thank you for your feedback, i truly appreciate the time you took to explain everything so clearly :pray:
I have a small remark :

I attempted to run the job locally using the modification you suggested. However, i encountered an issue.

My understanding is that the job should exit with code `0` if no unwanted tags are found. Since the command is prefixed with `!`, i expected it to invert the exit code, meaning success when no matches are found. Instead, i received an failure :thinking: :

With :
```
      - name: Ensure no debug tags are committed in feature files
        run: ! grep -rEn --include='*.feature' '@(testing|debug)' .
```
I got this result :

```
[Lint & Test/check-no-unwanted-strings] ⭐ Run Main Check unwanted tags
[Lint & Test/check-no-unwanted-strings]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/1] user= workdir=
[Lint & Test/check-no-unwanted-strings]   ❌  Failure - Main Check unwanted tags
[Lint & Test/check-no-unwanted-strings] exitcode '1': failure
```
When using `" "` or `|` it seems to work as intended
```
      - name: Ensure no debug tags are committed in feature files
        run: "! grep -rEn --include='*.feature' '@(testing|debug)' ."
```
Or this:
```
      - name: Ensure no debug tags are committed in feature files
        run: |
          ! grep -rEn --include='*.feature' '@(testing|debug)' .
```
Result : 
```
[Lint & Test/check-no-unwanted-strings]   ✅  Success - Main Checkout code
[Lint & Test/check-no-unwanted-strings] ⭐ Run Main Check unwanted tags
[Lint & Test/check-no-unwanted-strings]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/1] user= workdir=
[Lint & Test/check-no-unwanted-strings]   ✅  Success - Main Check unwanted tags
```

Let me know if i misunderstand anything. Thanks again for your help !

---

Oh yes, good catch! I didn't check myself but `!` must be a reserved YAML character. You can wrap it in double quotes (`"`) then 👍🏻

Have you tried running it on the repository with such tag in a feature file? I'm not 100% sure GH would prevent `smoke-test` from running based on the exit code (I believe but I'm not sure)?

---

Oh and by the way, do you get the output you shared from running the workflow locally? I didn't know one could do that 😯

---

> Oh yes, good catch! I didn't check myself but ! must be a reserved YAML character. You can wrap it in double quotes (") then 👍🏻

As far as i understand it is used to define tag when we want to use explicit value for example :
```
valeur: !!str 123   # interpred as a string
```
[source](https://en.wikipedia.org/wiki/YAML#Advanced_components)

> Have you tried running it on the repository with such tag in a feature file? I'm not 100% sure GH would prevent smoke-test from running based on the exit code (I believe but I'm not sure)?

I'm not sure if i fully understand the question :confused:
If the question is : did i run the **full workflow** locally to make sure the error code is preventing `smoke-test` from running, the answer is yes and no 😅
Indeed, i did the test by removing the `lint-rust` and `lint-openai` jobs _because these jobs fail locally, and i haven't taken the time to diagnose why._
So, if we consider test jobs such as:
```
jobs:
  check-no-unwanted-strings:
    runs-on: ubuntu-22.04
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Check unwanted tags
        run: "! grep -rEn --include='*.feature' '@(init|prosody)' ."

  smoke-test:
    if: github.event.pull_request.draft == false
    needs: [check-no-unwanted-strings]
[...]
```
By **voluntary setting tags present in the repository** and leaving the `smoke-test` dependency on the `check-no-unwanted-strings` job, we notice that `smoke-test` is not launched when the check fails.

```
# Listing all jobs
[antoine@fedora prose-pod-api (ci/no-testing-in-features)]$ act -l
Stage  Job ID                     Job name                   Workflow name   Workflow file  Events           
0      create-github-release      create-github-release      Release & Ship  release.yaml   workflow_run,push
0      ship-docker-image          ship-docker-image          Release & Ship  release.yaml   workflow_run,push
0      check-no-unwanted-strings  check-no-unwanted-strings  Lint & Test     test.yaml      push,pull_request
1      smoke-test                 smoke-test                 Lint & Test     test.yaml      pull_request,push
2      integration-test           integration-test           Lint & Test     test.yaml      push,pull_request
---
# Try to run the `smoke-test` job
[antoine@fedora prose-pod-api (ci/no-testing-in-features)]$ act -j smoke-test
[Lint & Test/check-no-unwanted-strings] ⭐ Run Set up job
[...]
[Lint & Test/check-no-unwanted-strings]   ✅  Success - Set up job
[Lint & Test/check-no-unwanted-strings] ⭐ Run Main Checkout code
[...]
[Lint & Test/check-no-unwanted-strings]   ✅  Success - Main Checkout code
[Lint & Test/check-no-unwanted-strings] ⭐ Run Main Ensure no debug tags are committed in feature files
[Lint & Test/check-no-unwanted-strings]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/1] user= workdir=
| ./features/init/init-first-account.feature:1:@init-first-account @init
| ./features/init/init-server-config.feature:1:@init-server-config @init
| ./features/init/init-workspace.feature:1:@init-workspace @init
| ./src/service/tests/features/prosody-config.feature:1:@prosody
[Lint & Test/check-no-unwanted-strings]   ❌  Failure - Main Ensure no debug tags are committed in feature files
[Lint & Test/check-no-unwanted-strings] exitcode '1': failure
[Lint & Test/check-no-unwanted-strings] ⭐ Run Complete job
[Lint & Test/check-no-unwanted-strings]   ✅  Success - Complete job
[Lint & Test/check-no-unwanted-strings] 🏁  Job failed
Error: Job 'check-no-unwanted-strings' failed
[antoine@fedora prose-pod-api (ci/no-testing-in-features)]$ 
```
I hope this answers your question.

> Oh and by the way, do you get the output you shared from running the workflow locally? I didn't know one could do that 😯

I actually discovered this tool recently : https://github.com/nektos/act

---

Thank you for the follow-up. Yes you answered my question 😉

Linting jobs might not work if you don't have the necessary tools installed, but it's fine if you just disabled them 👍🏻

---

> Linting jobs might not work if you don't have the necessary tools installed, but it's fine if you just disabled them 👍🏻

Yes, i think too, i will look at it :+1:

Since, i did the commit to correct the `grep` command in 9471b23 i will resolve this conversation